### ISSUE TITLE
# line 94 in lstm.py, return [tf.nn.rnn_cell.BasicLSTMCell(steps, state_is_tuple=True) for steps in layers]

### ISSUE THREAD

**Issue by bhupendra36** [mouradmourafiq/tensorflow-lstm-regression#3]
Input to BasicLSTMCell is number of hidden unit not time steps. As far as i know they are both independent parameters.
Can someone elaborate the things if i am wrong?
@mouradmourafiq 
@lepisma

**Comment by mouradmourafiq** [comment]
@bhupendra36 yes the `BasicLSTMCell` expects `num_units` as an argument, `time_steps` is just the name of the variable I created. I am not sure what are you confused about, the number of units are the the hidden units that define the time/term/sequence dependencies.

**Comment by bhupendra36** [comment]
@mouradmourafiq  , num_steps is used for Backpropagation through time ( defining the time dependencies). number of units increases the size of the layer.
Is there any specific reason for making `num_steps` and `num_units` equal ?

**Comment by mouradmourafiq** [comment]
@bhupendra36 `num_units` is the size for rnn cell which are used for the backpropagation through time. Each layer is made out of an `BasicLSTMCell`. This is my understanding of the [code](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/ops/rnn_cell.py#L283)

I think this is your source of confusion, there's no `num_steps` in the tensorflow code, I just named this variable to make it clearer.

Also currently there's no documentation for the `rnn`, I think it will be made public API in the next release.

You can also check other examples of rnn architecture used for NLP tasks, the `num_units ` is used as the size of the word embeddings.

**Comment by bhupendra36** [comment]
@mouradmourafiq , From tensorflow website In  [RNN tutorial](https://www.tensorflow.org/versions/r0.9/tutorials/recurrent/index.html) 

> ### Truncated Backpropagation
> In order to make the learning process tractable, it is a common practice to truncate the gradients for backpropagation to a fixed number (num_steps) of unrolled steps. This is easy to implement by feeding inputs of length num_steps at a time and doing backward pass after each iteration.

Time dependency can be taken in a way that how many time steps back RNN look while making a prediction. ( At each time step there will be input corresponding to that )

From this block of your code 
` X, y = generate_data(np.sin, np.linspace(0, 100, 10000), TIMESTEPS, seperate=False) `

RNN is given TIMESTEPS previous number of inputs to make the prediction
So one can easily infer that ` num_steps ( as mentioned in RNN tutorial )== TIMESTEPS `

So , what i was saying that `num_units` (number of such LSTM cell in a layer )are totally different and can be assigned different value than `TIMESTEPS` which is not done here
>regressor = learn.TensorFlowEstimator(model_fn=lstm_model(`TIMESTEPS`, RNN_LAYERS, DENSE_LAYERS), n_classes=0,
                                      verbose=2,  steps=TRAINING_STEPS, optimizer='Adagrad',
                                      learning_rate=0.03, batch_size=BATCH_SIZE)

**Comment by mouradmourafiq** [comment]
Exactly this is truncated backpropagation, which is optional and you can tweak/make it an option for learning in the model if you want. 

Truncating backpropagation is connected with the learning and not with the length of the sequence, and this parameter can't be bigger than length of the sequence.

From this [paper](http://arxiv.org/pdf/1308.0850v5.pdf)

>  This form of truncated backpropagation has been considered before for RNN language modelling [23], and found to speed up training (by reducing the sequence length and hence increasing the frequency of stochastic weight updates) without affecting the network’s ability to learn long-range dependencies.

But thanks for pointing it out, let me know if you try to use it and how does affect the learning in that case.

**Comment by bhupendra36** [comment]
Thanks for the explanation.
